### PR TITLE
DAOS-5896 obj: fix IOM handling for EC degraded fetch

### PR DESCRIPTION
**Pull request by liuxuezhao**
For EC obj degraded fetch -
1. need to merge the to-be-recovered recx list to the IOM,
2. need not to merge to IOM for data recovery task.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Review by wangdi1 — APPROVED** [review]
sorry, this is hard to follow. really need add some test case to verify.

File: src/object/cli_shard.c [hunk 1]
```diff
@@ -441,6 +453,22 @@ obj_ec_iom_merge(struct obj_reasb_req *reasb_req, uint32_t tgt_idx,
...
+		for (i = 0; i < recov_list->re_nr; i++) {
+			rc = iom_recx_merge(dst,
+					    &recov_list->re_items[i].re_recx,
+					    reasb_req->orr_iom_realloc);
+			if (rc == -DER_NOMEM)
+				break;
+			if (rc == -DER_REC2BIG) {
+				if (done)
+					dst->iom_nr_out += recov_list->re_nr;
+				rc = 0;
```
[the remark this answers is not in the document]
Comment (liuxuezhao) [reply src/object/cli_shard.c:467]: The original API defined as if user provide recx array not enough to hold it, then use the iom_nr_out to indicate how much buffered needed and don't return err code if no other failures. it seems not bad so I did not change it.

[the remark this answers is not in the document]
Comment (wangdi1) [reply src/object/cli_shard.c:467]: ah, I see. thanks.

**Review by wangdi1 — DISMISSED** [review]

**Review by wangdi1 — APPROVED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Review by wangdi1 — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]

File: src/client/array/dc_array.c [hunk 2]
```diff
@@ -1099,6 +1099,7 @@ struct hole_params {
 static int
 zero_out_cb(uint8_t *buf, size_t len, void *args)
 {
+	D_DEBUG(DB_IO, "zero hole segment, buf %p, len %zu\n", buf, len);
```
[the remark this answers is not in the document]
Comment (liuxuezhao) [reply src/client/array/dc_array.c:1102]: right, but silently zero user buffer cause hard to trace for failure case. (I spent quite some time to finger out the issue).
we can change later if find the log is too noisy, maybe change to DB_TRACE.
